src/action_dispatcher.rs

```diff
@@ -25,12 +26,42 @@ impl ActionDispatcher {
     }
 
     // Execute Actions created by EventHandler. This should be the only public method of ActionDispatcher.
-    pub fn on_action(&mut self, action: Action) -> anyhow::Result<()> {
-        match action {
-            Action::KeyEvent(key_event) => self.on_key_event(key_event)?,
-            Action::InputEvent(event) => self.send_event(event)?,
-            Action::Command(command) => self.run_command(command),
-            Action::Delay(duration) => thread::sleep(duration),
+    pub fn on_action(&mut self, actions: Vec<Action>) -> anyhow::Result<()> {
```

Comment:
You should be able to emit a vector of `MouseMovementEvent` as a type of Action and handle it here instead of changing the argument type to `actions: Vec<Action>`. The idea of `EventHandler` / `ActionDispatcher` is to isolate the portion that has side effects and is harder to test (`ActionDispatcher`) from the (basically side-effect-free) pure function (`EventHandler`). Every complicated event transformation should be done in `EventHandler` so it can be tested, and `ActionDispatcher` should be dumb enough to obviate tests.